src/parser/expr.rs

```diff
@@ -1,116 +1,103 @@
-        ])? {
-            let right = self.sum()?;
-            let span = self.spans.merge(&left, &right);
-            return Ok(ExprKind::BinOp(op, self.alloc(left), self.alloc(right)).with_span(span));
-        } else {
-            Ok(left)
-        }
+        Ok(lhs)
     }
 
-    // sum = sum '+' term
-    //     | sum '-' term
-    //     | term
-    fn sum(&mut self) -> ParseResult<Expr<'s, 'a>> {
-        let mut left = self.term()?;
-        while let Some((op, _span)) =
-            self.accept_any([(tok![+], BinOp::Add), (tok![-], BinOp::Sub)])?
-        {
-            let right = self.term()?;
-            let span = self.spans.merge(&left, &right);
-            left = ExprKind::BinOp(op, self.alloc(left), self.alloc(right)).with_span(span);
-        }
-        Ok(left)
+    fn binary_expr_rhs(
+        &mut self,
+        lhs: Expr<'s, 'a>,
+        operator: BinaryOp,
+    ) -> Result<Expr<'s, 'a>, ParseError> {
+        // we peeked already that an operator is coming, we just need to know its span and progress the parser
+        let (_, span) = self.next()?;
+        let span = self.spans.store(span);
+        let operator = operator.with_span(span);
+
+        // if we continue,
+        // parse the left hand side of the next expression
+        let new_lhs = self.unary()?;
+        // and now parse a next expression, with this newly parsed
+        // left hand side. Maybe, it parses an operator too many,
+        // in that case it returns it and we cache it in `maybe_operator`
+        // and use it next if we determine we should not return it ourselves
+        let rhs = self.binary_expr(new_lhs, &operator)?;
+
+        // make an expression, and loop
+        let span = self.spans.merge(&lhs, &rhs);
+
+        Ok(ExprKind::BinaryOp(operator, self.alloc(lhs), self.alloc(rhs)).with_span(span))
     }
 
-    // term = term '*' factor
-    //      | term '/' factor
-    //      | factor
-    fn term(&mut self) -> ParseResult<Expr<'s, 'a>> {
-        let mut left = self.factor()?;
-        while let Some((op, _span)) =
-            self.accept_any([(tok![*], BinOp::Mul), (tok![/], BinOp::Div)])?
-        {
-            let right = self.factor()?;
-            let span = self.spans.merge(&left, &right);
-            left = ExprKind::BinOp(op, self.alloc(left), self.alloc(right)).with_span(span);
-        }
-        Ok(left)
+    fn peek_binary_operator(&mut self) -> ParseResult<Option<BinaryOp>> {
+        self.peek_any([
+            (tok![*], BinaryOp::Mul),
+            (tok![/], BinaryOp::Div),
+            (tok![+], BinaryOp::Add),
+            (tok![-], BinaryOp::Sub),
+            (tok![&&], BinaryOp::LogicalAnd),
+            (tok![||], BinaryOp::LogicalOr),
+            (tok![>], BinaryOp::Gt),
+            (tok![>=], BinaryOp::Gte),
+            (tok![<], BinaryOp::Lt),
+            (tok![<=], BinaryOp::Lte),
+            (tok![==], BinaryOp::Eq),
+            (tok![!=], BinaryOp::Neq),
```

Comment:
Would it be possible to refactor this array out to `BinaryOp`?  It just needs to provide a `const ALL: [Self]` with every possible operator, and a `fn as_token(&self) -> Token` getter.  Though in that case I think `peek_any` would have to take an iterator as input...

Reply:
since we only do it here, and probably only ever want it here I don't see the benefit
we can always do that when the time comes